### PR TITLE
[Fix] PARSeq pytorch fixes

### PR DESCRIPTION
**Pull request by felixdittrich92**
This PR:

- contains fixes for the pytorch version of PARSeq

- [x] Fix masks for training

dummy run:

Validation loss decreased 0.407558 --> 0.322983: saving state...
Epoch 8/20 - Validation loss: 0.322983 (Exact: 82.78% | Partial: 83.81%)

### CODE REVIEW

**Comment by felixdittrich92** [comment]
@baudm

train:

```
tgt_perms: tensor([[0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
        [0, 9, 8, 7, 6, 5, 4, 3, 2, 1],
        [0, 6, 5, 2, 3, 4, 7, 1, 8, 9],
        [0, 8, 1, 7, 4, 3, 2, 5, 6, 9],
        [0, 5, 6, 3, 4, 8, 1, 7, 2, 9],
        [0, 2, 7, 1, 8, 4, 3, 6, 5, 9]], device='cuda:0', dtype=torch.int32)
padding_mask: tensor([[True, True, True, True, True, True, True, True, True]],
       device='cuda:0')
padding_mask.shape: torch.Size([1, 9])
key_padding_mask_expanded: tensor([[[[True, True, True, True, True, True, True, True, True]]]],
       device='cuda:0')
key_padding_mask_expanded.shape: torch.Size([1, 1, 1, 9])
target_mask: tensor([[1, 0, 0, 0, 0, 0, 0, 0, 0],
        [1, 1, 0, 0, 0, 0, 0, 0, 0],
        [1, 1, 1, 0, 0, 0, 0, 0, 0],
        [1, 1, 1, 1, 0, 0, 0, 0, 0],
        [1, 1, 1, 1, 1, 0, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1]], device='cuda:0', dtype=torch.int32)
target_mask.shape: torch.Size([9, 9])
attn_mask_expanded: tensor([[[[ True, False, False, False, False, False, False, False, False],
          [ True,  True, False, False, False, False, False, False, False],
          [ True,  True,  True, False, False, False, False, False, False],
          [ True,  True,  True,  True, False, False, False, False, False],
          [ True,  True,  True,  True,  True, False, False, False, False],
          [ True,  True,  True,  True,  True,  True, False, False, False],
          [ True,  True,  True,  True,  True,  True,  True, False, False],
          [ True,  True,  True,  True,  True,  True,  True,  True, False],
          [ True,  True,  True,  True,  True,  True,  True,  True,  True]]]],
       device='cuda:0')
attn_mask_expanded.shape: torch.Size([1, 1, 9, 9])
mask: tensor([[[[1, 0, 0, 0, 0, 0, 0, 0, 0],
          [1, 1, 0, 0, 0, 0, 0, 0, 0],
          [1, 1, 1, 0, 0, 0, 0, 0, 0],
          [1, 1, 1, 1, 0, 0, 0, 0, 0],
          [1, 1, 1, 1, 1, 0, 0, 0, 0],
          [1, 1, 1, 1, 1, 1, 0, 0, 0],
          [1, 1, 1, 1, 1, 1, 1, 0, 0],
          [1, 1, 1, 1, 1, 1, 1, 1, 0],
          [1, 1, 1, 1, 1, 1, 1, 1, 1]]]], device='cuda:0', dtype=torch.int32)
mask.shape: torch.Size([1, 1, 9, 9])
target_mask: tensor([[1, 0, 1, 1, 1, 1, 1, 1, 1],
        [1, 0, 0, 1, 1, 1, 1, 1, 1],
        [1, 0, 0, 0, 1, 1, 1, 1, 1],
        [1, 0, 0, 0, 0, 1, 1, 1, 1],
        [1, 0, 0, 0, 0, 0, 1, 1, 1],
        [1, 0, 0, 0, 0, 0, 0, 1, 1],
        [1, 0, 0, 0, 0, 0, 0, 0, 1],
        [1, 0, 0, 0, 0, 0, 0, 0, 0],
        [1, 0, 0, 0, 0, 0, 0, 0, 0]], device='cuda:0', dtype=torch.int32)
target_mask.shape: torch.Size([9, 9])
attn_mask_expanded: tensor([[[[ True, False,  True,  True,  True,  True,  True,  True,  True],
          [ True, False, False,  True,  True,  True,  True,  True,  True],
          [ True, False, False, False,  True,  True,  True,  True,  True],
          [ True, False, False, False, False,  True,  True,  True,  True],
          [ True, False, False, False, False, False,  True,  True,  True],
          [ True, False, False, False, False, False, False,  True,  True],
          [ True, False, False, False, False, False, False, False,  True],
          [ True, False, False, False, False, False, False, False, False],
          [ True, False, False, False, False, False, False, False, False]]]],
       device='cuda:0')
attn_mask_expanded.shape: torch.Size([1, 1, 9, 9])
mask: tensor([[[[1, 0, 1, 1, 1, 1, 1, 1, 1],
          [1, 0, 0, 1, 1, 1, 1, 1, 1],
          [1, 0, 0, 0, 1, 1, 1, 1, 1],
          [1, 0, 0, 0, 0, 1, 1, 1, 1],
          [1, 0, 0, 0, 0, 0, 1, 1, 1],
          [1, 0, 0, 0, 0, 0, 0, 1, 1],
          [1, 0, 0, 0, 0, 0, 0, 0, 1],
          [1, 0, 0, 0, 0, 0, 0, 0, 0],
          [1, 0, 0, 0, 0, 0, 0, 0, 0]]]], device='cuda:0', dtype=torch.int32)
mask.shape: torch.Size([1, 1, 9, 9])
target_mask: tensor([[1, 0, 1, 1, 1, 1, 1, 1, 0],
        [1, 0, 0, 0, 0, 1, 1, 0, 0],
        [1, 0, 1, 0, 0, 1, 1, 0, 0],
        [1, 0, 1, 1, 0, 1, 1, 0, 0],
        [1, 0, 0, 0, 0, 0, 1, 0, 0],
        [1, 0, 0, 0, 0, 0, 0, 0, 0],
        [1, 0, 1, 1, 1, 1, 1, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1]], device='cuda:0', dtype=torch.int32)
target_mask.shape: torch.Size([9, 9])
attn_mask_expanded: tensor([[[[ True, False,  True,  True,  True,  True,  True,  True, False],
          [ True, False, False, False, False,  True,  True, False, False],
          [ True, False,  True, False, False,  True,  True, False, False],
          [ True, False,  True,  True, False,  True,  True, False, False],
          [ True, False, False, False, False, False,  True, False, False],
          [ True, False, False, False, False, False, False, False, False],
          [ True, False,  True,  True,  True,  True,  True, False, False],
          [ True,  True,  True,  True,  True,  True,  True,  True, False],
          [ True,  True,  True,  True,  True,  True,  True,  True,  True]]]],
       device='cuda:0')
attn_mask_expanded.shape: torch.Size([1, 1, 9, 9])
mask: tensor([[[[1, 0, 1, 1, 1, 1, 1, 1, 0],
          [1, 0, 0, 0, 0, 1, 1, 0, 0],
          [1, 0, 1, 0, 0, 1, 1, 0, 0],
          [1, 0, 1, 1, 0, 1, 1, 0, 0],
          [1, 0, 0, 0, 0, 0, 1, 0, 0],
          [1, 0, 0, 0, 0, 0, 0, 0, 0],
          [1, 0, 1, 1, 1, 1, 1, 0, 0],
          [1, 1, 1, 1, 1, 1, 1, 1, 0],
          [1, 1, 1, 1, 1, 1, 1, 1, 1]]]], device='cuda:0', dtype=torch.int32)
mask.shape: torch.Size([1, 1, 9, 9])
target_mask: tensor([[1, 0, 0, 0, 0, 0, 0, 0, 1],
        [1, 1, 0, 1, 1, 0, 0, 1, 1],
        [1, 1, 0, 0, 1, 0, 0, 1, 1],
        [1, 1, 0, 0, 0, 0, 0, 1, 1],
        [1, 1, 1, 1, 1, 0, 0, 1, 1],
        [1, 1, 1, 1, 1, 1, 0, 1, 1],
        [1, 1, 0, 0, 0, 0, 0, 0, 1],
        [1, 0, 0, 0, 0, 0, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1]], device='cuda:0', dtype=torch.int32)
target_mask.shape: torch.Size([9, 9])
attn_mask_expanded: tensor([[[[ True, False, False, False, False, False, False, False,  True],
          [ True,  True, False,  True,  True, False, False,  True,  True],
          [ True,  True, False, False,  True, False, False,  True,  True],
          [ True,  True, False, False, False, False, False,  True,  True],
          [ True,  True,  True,  True,  True, False, False,  True,  True],
          [ True,  True,  True,  True,  True,  True, False,  True,  True],
          [ True,  True, False, False, False, False, False, False,  True],
          [ True, False, False, False, False, False, False, False, False],
          [ True,  True,  True,  True,  True,  True,  True,  True,  True]]]],
       device='cuda:0')
attn_mask_expanded.shape: torch.Size([1, 1, 9, 9])
mask: tensor([[[[1, 0, 0, 0, 0, 0, 0, 0, 1],
          [1, 1, 0, 1, 1, 0, 0, 1, 1],
          [1, 1, 0, 0, 1, 0, 0, 1, 1],
          [1, 1, 0, 0, 0, 0, 0, 1, 1],
          [1, 1, 1, 1, 1, 0, 0, 1, 1],
          [1, 1, 1, 1, 1, 1, 0, 1, 1],
          [1, 1, 0, 0, 0, 0, 0, 0, 1],
          [1, 0, 0, 0, 0, 0, 0, 0, 0],
          [1, 1, 1, 1, 1, 1, 1, 1, 1]]]], device='cuda:0', dtype=torch.int32)
mask.shape: torch.Size([1, 1, 9, 9])
target_mask: tensor([[1, 0, 0, 1, 1, 1, 1, 0, 1],
        [1, 1, 0, 1, 1, 1, 1, 1, 1],
        [1, 0, 0, 0, 0, 1, 1, 0, 0],
        [1, 0, 0, 1, 0, 1, 1, 0, 0],
        [1, 0, 0, 0, 0, 0, 0, 0, 0],
        [1, 0, 0, 0, 0, 1, 0, 0, 0],
        [1, 1, 0, 1, 1, 1, 1, 0, 1],
        [1, 0, 0, 1, 1, 1, 1, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1]], device='cuda:0', dtype=torch.int32)
target_mask.shape: torch.Size([9, 9])
attn_mask_expanded: tensor([[[[ True, False, False,  True,  True,  True,  True, False,  True],
          [ True,  True, False,  True,  True,  True,  True,  True,  True],
          [ True, False, False, False, False,  True,  True, False, False],
          [ True, False, False,  True, False,  True,  True, False, False],
          [ True, False, False, False, False, False, False, False, False],
          [ True, False, False, False, False,  True, False, False, False],
          [ True,  True, False,  True,  True,  True,  True, False,  True],
          [ True, False, False,  True,  True,  True,  True, False, False],
          [ True,  True,  True,  True,  True,  True,  True,  True,  True]]]],
       device='cuda:0')
attn_mask_expanded.shape: torch.Size([1, 1, 9, 9])
mask: tensor([[[[1, 0, 0, 1, 1, 1, 1, 0, 1],
          [1, 1, 0, 1, 1, 1, 1, 1, 1],
          [1, 0, 0, 0, 0, 1, 1, 0, 0],
          [1, 0, 0, 1, 0, 1, 1, 0, 0],
          [1, 0, 0, 0, 0, 0, 0, 0, 0],
          [1, 0, 0, 0, 0, 1, 0, 0, 0],
          [1, 1, 0, 1, 1, 1, 1, 0, 1],
          [1, 0, 0, 1, 1, 1, 1, 0, 0],
          [1, 1, 1, 1, 1, 1, 1, 1, 1]]]], device='cuda:0', dtype=torch.int32)
mask.shape: torch.Size([1, 1, 9, 9])
target_mask: tensor([[1, 0, 1, 0, 0, 0, 0, 1, 0],
        [1, 0, 0, 0, 0, 0, 0, 0, 0],
        [1, 1, 1, 0, 1, 0, 0, 1, 1],
        [1, 1, 1, 0, 0, 0, 0, 1, 1],
        [1, 1, 1, 1, 1, 0, 1, 1, 1],
        [1, 1, 1, 1, 1, 0, 0, 1, 1],
        [1, 0, 1, 0, 0, 0, 0, 0, 0],
        [1, 1, 1, 0, 0, 0, 0, 1, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1]], device='cuda:0', dtype=torch.int32)
target_mask.shape: torch.Size([9, 9])
attn_mask_expanded: tensor([[[[ True, False,  True, False, False, False, False,  True, False],
          [ True, False, False, False, False, False, False, False, False],
          [ True,  True,  True, False,  True, False, False,  True,  True],
          [ True,  True,  True, False, False, False, False,  True,  True],
          [ True,  True,  True,  True,  True, False,  True,  True,  True],
          [ True,  True,  True,  True,  True, False, False,  True,  True],
          [ True, False,  True, False, False, False, False, False, False],
          [ True,  True,  True, False, False, False, False,  True, False],
          [ True,  True,  True,  True,  True,  True,  True,  True,  True]]]],
       device='cuda:0')
attn_mask_expanded.shape: torch.Size([1, 1, 9, 9])
mask: tensor([[[[1, 0, 1, 0, 0, 0, 0, 1, 0],
          [1, 0, 0, 0, 0, 0, 0, 0, 0],
          [1, 1, 1, 0, 1, 0, 0, 1, 1],
          [1, 1, 1, 0, 0, 0, 0, 1, 1],
          [1, 1, 1, 1, 1, 0, 1, 1, 1],
          [1, 1, 1, 1, 1, 0, 0, 1, 1],
          [1, 0, 1, 0, 0, 0, 0, 0, 0],
          [1, 1, 1, 0, 0, 0, 0, 1, 0],
          [1, 1, 1, 1, 1, 1, 1, 1, 1]]]], device='cuda:0', dtype=torch.int32)
mask.shape: torch.Size([1, 1, 9, 9])
```

**Comment by felixdittrich92** [comment]
val decode_autoregressive:

```
query_mask: tensor([[1, 0, 0],
        [1, 1, 0],
        [1, 1, 1]], device='cuda:0', dtype=torch.int32)
query_mask.shape: torch.Size([3, 3])
updated query_mask: tensor([[1, 0, 1],
        [1, 1, 0],
        [1, 1, 1]], device='cuda:0', dtype=torch.int32)
updated query_mask.shape: torch.Size([3, 3])
target_pad_mask: tensor([[[[ True, False, False]]]], device='cuda:0')
target_pad_mask.shape: torch.Size([1, 1, 1, 3])
mask: tensor([[[[1, 0, 0],
          [1, 0, 0],
          [1, 0, 0]]]], device='cuda:0', dtype=torch.int32)
mask.shape: torch.Size([1, 1, 3, 3])
```

**Review by charlesmindee — APPROVED** [review]
Thanks for that, LGTM
